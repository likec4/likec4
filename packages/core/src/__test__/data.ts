import { indexBy } from 'remeda'
import { ModelIndex } from '../model-index'
import type { Element, ElementKind, Fqn, Relation, RelationID, Tag } from '../types'

/**
              ┌──────────────────────────────────────────────────┐
              │                      cloud                       │
              │  ┌───────────────────────────────────────────┐   │
              │  │                 frontend                  │   │
┏━━━━━━━━━━┓  │  │   ┏━━━━━━━━━━━━━┓   ┏━━━━━━━━━━━━━━━━┓    │   │   ┏━━━━━━━━━━━┓
┃          ┃  │  │   ┃             ┃   ┃                ┃    │   │   ┃           ┃
┃ customer ┃──┼──┼──▶┃  dashboard  ┃   ┃   adminpanel   ┃◀───┼───┼───┃  support  ┃
┃          ┃  │  │   ┃             ┃   ┃                ┃    │   │   ┃           ┃
┗━━━━━━━━━━┛  │  │   ┗━━━━━━┳━━━━━━┛   ┗━━━━━━━━┳━━━━━━━┛    │   │   ┗━━━━━━━━━━━┛
              │  └──────────┼───────────────────┼────────────┘   │
              │             ├───────────────────┘                │
              │             │                                    │
              │  ┌──────────┼────────────────────────────────┐   │
              │  │          ▼       backend                  │   │
              │  │   ┏━━━━━━━━━━━━━┓       ┏━━━━━━━━━━━━━┓   │   │
              │  │   ┃             ┃       ┃             ┃   │   │
              │  │   ┃  graphlql   ┃──────▶┃   storage   ┃   │   │
              │  │   ┃             ┃       ┃             ┃   │   │
              │  │   ┗━━━━━━━━━━━━━┛       ┗━━━━━━┳━━━━━━┛   │   │
              │  └────────────────────────────────┼──────────┘   │
              └───────────────────────────────────┼──────────────┘
                                                  │
                                        ┌─────────┼─────────┐
                                        │ amazon  │         │
                                        │         ▼         │
                                        │ ┏━━━━━━━━━━━━━━┓  │
                                        │ ┃              ┃  │
                                        │ ┃      s3      ┃  │
                                        │ ┃              ┃  │
                                        │ ┗━━━━━━━━━━━━━━┛  │
                                        └───────────────────┘

specification {
  element actor
  element system
  element container
  element component

  tag old
}

model {

  actor customer
  actor support

  system cloud {
    container backend {
      component graphql
      component storage {
        #old
      }

      graphql -> storage
    }

    container frontend {
      component dashboard {
        -> graphql
      }
      component adminPanel {
        #old
        -> graphql
      }
    }
  }

  customer -> dashboard
  support -> adminPanel

  system amazon {
    component s3

    cloud.backend.storage -> s3
  }

}

 */
const el = ({
  id,
  kind,
  title,
  ...props
}: Partial<Omit<Element, 'id' | 'kind'>> & { id: string; kind: string }): Element => ({
  id: id as Fqn,
  kind: kind as ElementKind,
  title: title ?? id,
  description: null,
  technology: null,
  tags: null,
  links: null,
  ...props
})

export const fakeElements = {
  amazon: el({
    id: 'amazon',
    kind: 'system',
    title: 'amazon'
  }),
  cloud: el({
    id: 'cloud',
    kind: 'system',
    title: 'cloud'
  }),
  customer: el({
    id: 'customer',
    kind: 'actor',
    title: 'customer',
    shape: 'person'
  }),
  support: el({
    id: 'support',
    kind: 'actor',
    title: 'support',
    shape: 'person'
  }),
  'amazon.s3': el({
    id: 'amazon.s3',
    kind: 'component',
    title: 's3',
    shape: 'storage'
  }),
  'cloud.backend': el({
    id: 'cloud.backend',
    kind: 'container',
    title: 'backend'
  }),
  'cloud.frontend': el({
    id: 'cloud.frontend',
    kind: 'container',
    title: 'frontend',
    shape: 'browser'
  }),
  'cloud.backend.graphql': el({
    id: 'cloud.backend.graphql',
    kind: 'component',
    title: 'graphql'
  }),
  'cloud.backend.storage': el({
    id: 'cloud.backend.storage',
    kind: 'component',
    title: 'storage',
    tags: ['old' as Tag]
  }),
  'cloud.frontend.adminPanel': el({
    id: 'cloud.frontend.adminPanel',
    kind: 'component',
    title: 'adminPanel',
    tags: ['old' as Tag]
  }),
  'cloud.frontend.dashboard': el({
    id: 'cloud.frontend.dashboard',
    kind: 'component',
    title: 'dashboard'
  })
} satisfies Record<string, Element>

export type FakeElementIds = keyof typeof fakeElements

const rel = ({
  source,
  target,
  title
}: {
  source: FakeElementIds
  target: FakeElementIds
  title?: string
}): Relation => ({
  id: `${source}:${target}` as RelationID,
  title: title ?? '',
  source: source as Fqn,
  target: target as Fqn
})

export const fakeRelations = [
  rel({
    source: 'customer',
    target: 'cloud.frontend.dashboard'
  }),
  rel({
    source: 'support',
    target: 'cloud.frontend.adminPanel'
  }),
  rel({
    source: 'cloud.backend.storage',
    target: 'amazon.s3'
  }),
  rel({
    source: 'cloud.backend.graphql',
    target: 'cloud.backend.storage'
  }),
  rel({
    source: 'cloud.frontend.dashboard',
    target: 'cloud.backend.graphql'
  }),
  rel({
    source: 'cloud.frontend.adminPanel',
    target: 'cloud.backend.graphql'
  })
]

export type FakeRelationIds = keyof typeof fakeRelations

export const fakeModel = () =>
  ModelIndex.from({
    elements: fakeElements,
    relations: indexBy(fakeRelations, r => r.id)
  })
