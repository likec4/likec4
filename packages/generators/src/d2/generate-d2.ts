import type { ComputedEdge, ComputedNode, ComputedView, NodeId } from '@likec4/core/types'
import { CompositeGeneratorNode, NL, joinToNode, toString } from 'langium/lib/generator'
import { isNil } from 'rambdax'

const capitalizeFirstLetter = (value: string) =>
  value.charAt(0).toLocaleUpperCase() + value.slice(1)

const fqnName = (nodeId: string): string => nodeId.split('.').map(capitalizeFirstLetter).join('')

const nodeName = (node: ComputedNode): string => {
  return fqnName(node.parent ? node.id.slice(node.parent.length + 1) : node.id)
}

const d2direction = ({ autoLayout }: ComputedView) => {
  switch (autoLayout) {
    case "TB": { return "down" }
    case "BT": { return "up" }
    case "LR": { return "right" }
    case "RL": { return "left" }
  }
}

const d2shape = ({ shape }: ComputedNode) => {
  switch (shape) {
    case 'queue':
    case 'cylinder':
    case 'rectangle':
    case 'person': {
      return shape
    }
    case 'storage': {
      return 'stored_data' as const
    }
    case 'mobile':
    case 'browser': {
      return 'rectangle' as const
    }
  }
}

export function generateD2<V extends ComputedView>(view: V) {
  const { nodes, edges } = view
  const names = new Map<NodeId, string>()

  const printNode = (node: ComputedNode, parentName?: string): CompositeGeneratorNode => {
    const name = nodeName(node)
    const fqnName = (parentName ? parentName + '.' : '') + name
    names.set(node.id, fqnName)

    const label = node.title.replaceAll('\n', '\\n')
    const shape = d2shape(node)

    return new CompositeGeneratorNode()
      .append(name, ': {', NL)
      .indent({
        indentedChildren: indent =>
          indent
            .append('label: "', label, '"', NL)
            .appendIf(
              shape !== 'rectangle',
              'shape: ', shape, NL
            )
            .appendIf(
              node.children.length > 0,
              NL,
              joinToNode(
                nodes.filter(n => n.parent === node.id),
                n => printNode(n, fqnName)
              )
            ),
        indentation: 2
      })
      .append('}', NL)
  }
  //     return `${names.get(edge.source)} -> ${names.get(edge.target)}${edge.label ? ': ' + edge.label : ''}`
  const printEdge = (edge: ComputedEdge): CompositeGeneratorNode => {
    return new CompositeGeneratorNode()
      .append(names.get(edge.source), ' -> ', names.get(edge.target))
      .append(out => edge.label && out.append(': ', edge.label.replaceAll('\n', '\\n')))
  }

  return toString(
    new CompositeGeneratorNode()
      .append('direction: ', d2direction(view), NL, NL)
      .append(
        joinToNode(
          nodes.filter(n => isNil(n.parent)),
          n => printNode(n),
          {
            appendNewLineIfNotEmpty: true
          }
        )
      )
      .appendIf(
        edges.length > 0,
        NL,
        joinToNode(edges, e => printEdge(e), {
          appendNewLineIfNotEmpty: true
        })
      )
  )
}
