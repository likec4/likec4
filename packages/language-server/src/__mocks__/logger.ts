import { vi } from 'vitest'
import type { Logger } from '../logger'

export const logger = {
  debug: vi.fn(),
  info: vi.fn(),
  warn: vi.fn(),
  log: vi.fn(),
  error: vi.fn(),
  trace: vi.fn()
} satisfies Logger
