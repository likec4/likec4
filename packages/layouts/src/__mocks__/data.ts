import { computeView, assignNavigateTo } from '@likec4/core'
import { cloudView, cloud3levels, amazonView, fakeModel, indexView } from './model'

const model = fakeModel()

export const [
  computedIndexView,
  computedCloudView,
  computedCloud3levels,
  computedAmazonView
] = assignNavigateTo([
  computeView(indexView, model)!,
  computeView(cloudView, model)!,
  computeView(cloud3levels, model)!,
  computeView(amazonView, model)!
])
