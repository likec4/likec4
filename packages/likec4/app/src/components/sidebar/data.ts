import type { DiagramView } from '@likec4/core'
import type { TreeNodeData } from '@mantine/core'
import { useMemo } from 'react'
import { find, isTruthy, values } from 'remeda'
import { useLikeC4Views } from 'virtual:likec4/store'

interface DiagramTreeNodeData {
  label: string
  value: string
  type: 'file' | 'folder' | 'diagram'
  children: DiagramTreeNodeData[]
}

export const isTreeNodeData = (node: TreeNodeData): node is DiagramTreeNodeData =>
  'type' in node && ['file', 'folder', 'diagram'].includes(node.type as any)

function compareTreeNodes(a: DiagramTreeNodeData, b: DiagramTreeNodeData) {
  if (a.children.length === 0 && b.children.length > 0) {
    return 1
  }
  if (a.children.length > 0 && b.children.length === 0) {
    return -1
  }
  return a.label.localeCompare(b.label)
}

function buildDiagramTreeData(views: DiagramView[]): DiagramTreeNodeData[] {
  const root: DiagramTreeNodeData = {
    value: '',
    label: 'Diagrams',
    type: 'folder',
    children: []
  }

  const findParent = (path: string): DiagramTreeNodeData => {
    let parent = root
    if (path === '') {
      return parent
    }
    const segments = path.split('/')
    const traversed = ['@fs'] as string[]
    while (segments.length) {
      const label = segments.shift() as string
      traversed.push(label)
      const value = traversed.join('/')
      let node = find(parent.children!, n => n.value === value)
      if (!node) {
        node = { label, value, type: 'folder', children: [] }
        parent.children.push(node)
      }
      parent = node
    }
    return parent
  }

  for (const view of views) {
    const parent = findParent(view.relativePath ?? '')
    parent.children.push({
      value: view.id,
      label: view.title ?? view.id,
      type: 'diagram',
      children: []
    })
    if (parent !== root) {
      parent.children.sort(compareTreeNodes)
      if (isTruthy(view.relativePath) && parent.type !== 'file') {
        parent.type = 'file'
      }
    }
  }

  return root.children.sort(compareTreeNodes)
}

// const $diagramsTree = batched($views, views => buildDiagramTreeData(values(views)))

export function useDiagramsTreeData() {
  const views = useLikeC4Views() as unknown as Record<string, DiagramView>
  return useMemo(() => buildDiagramTreeData(values(views)), [views])
}
