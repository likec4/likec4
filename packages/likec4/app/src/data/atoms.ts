import type { ExtractAtomValue } from 'jotai'
import { atom } from 'jotai'
import { atomFamily, atomWithReducer, splitAtom } from 'jotai/utils'
import { equals, groupBy, mapObject, values } from 'rambdax'
import { LikeC4Views } from '~likec4'
import { buildDiagramTreeAtom } from './sidebar-diagram-tree'
import type { DiagramView } from '@likec4/core'

function atomWithCompare<Value>(initialValue: Value) {
  return atomWithReducer(initialValue, (prev: Value, next: Value) => {
    if (equals(prev, next)) {
      return prev
    }
    return next
  })
}

const likec4ViewFamily = atomFamily(
  (view: DiagramView) => {
    const atom = atomWithCompare(view)
    atom.debugLabel = `view#${view.id}`
    return atom
  },
  (a, b) => a.id === b.id
)

const mapToAtoms = (views: typeof LikeC4Views) => mapObject(view => likec4ViewFamily(view), views)

const _viewsAtom = atom(mapToAtoms(LikeC4Views))
_viewsAtom.debugLabel = '_views'

export const viewsAtom = atom(
  get => get(_viewsAtom),
  (_, set, update: typeof LikeC4Views) => {
    set(
      _viewsAtom,
      mapObject(view => {
        const atom = likec4ViewFamily(view)
        set(atom, view)
        return atom
      }, update)
    )
  }
)
viewsAtom.debugLabel = 'views'

const indexPageTilesAtom = atom(get => {
  const views = values(get(viewsAtom))
  const byPath = groupBy(v => get(v).relativePath ?? '', views)
  return Object.entries(byPath)
    .map(([path, views]) => ({
      path,
      isRoot: path === '',
      views
    }))
    .sort((a, b) => {
      return a.path.localeCompare(b.path)
    })
})
export type IndexPageTile = ExtractAtomValue<typeof indexPageTilesAtom>[number]

const byPath = (tile: IndexPageTile) => tile.path
export const indexPageTilesAtomsAtom = splitAtom(indexPageTilesAtom, byPath)

// export type DashboardTile = ExtractAtomValue<typeof dashboardTilesAtom>[number]

// const key = (tile: DashboardTile) => tile.path

// export const diagramsTreeAtom = atom(buildDiagramTreeAtom(LikeC4Views))
export const diagramsTreeAtom = atom(get => {
  const views = values(get(viewsAtom))
  return buildDiagramTreeAtom(views.map(v => get(v)))
})

export const selectLikeC4ViewAtom = (viewId: string) => {
  return atom(get => {
    const viewAtom = get(viewsAtom)[viewId]
    return viewAtom ? get(viewAtom) : null
  })
}

if (import.meta.hot) {
  let $updateViews: undefined | ((update: typeof LikeC4Views) => void)

  viewsAtom.onMount = set => {
    $updateViews = set
    return () => {
      $updateViews = undefined
    }
  }

  import.meta.hot.accept('/@vite-plugin-likec4/likec4-generated', md => {
    const update = md?.LikeC4Views as typeof LikeC4Views | undefined
    if (update) {
      $updateViews?.(update)
    }
  })
}
