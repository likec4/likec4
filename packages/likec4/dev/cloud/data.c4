model {

  extend cloud.data {

    hitsAnomalies = app 'Hits Anomalies' {
      description '
        Process page hits.
        Search for anomalies.
      '

      -> amazon.sqs.hits 'process hits data'
    }

  }

}

views {

  view cloud_data of cloud.data {
    title "[context] Data Services"
    include
        *,
        cloud,
        amazon,
        amazon.sqs,
        cloud.data -> amazon.sqs.*

    style cloud {
      color muted
    }
    style amazon.* {
      color secondary
    }
  }

}
