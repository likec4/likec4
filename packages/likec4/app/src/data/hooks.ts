import { useAtomValue } from 'jotai'
import { useMemo } from 'react'
import { diagramsTreeAtom, indexPageTilesAtomsAtom, selectLikeC4ViewAtom } from './atoms'

export const useLikeC4View = (viewId: string) => {
  const anAtom = useMemo(() => selectLikeC4ViewAtom(viewId), [viewId])
  return useAtomValue(anAtom)
}

export const useIndexPageTileAtoms = () => {
  return useAtomValue(indexPageTilesAtomsAtom)
}

export type { IndexPageTile } from './atoms'

export const useDiagramsTree = () => useAtomValue(diagramsTreeAtom)
