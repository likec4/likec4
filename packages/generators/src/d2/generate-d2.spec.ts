import { expect, test } from 'vitest'
import { fakeDiagram, fakeDiagram2, fakeComputedView3Levels } from '../__mocks__/data'
import { generateD2 } from './generate-d2'

test('generate D2 - fakeDiagram', () => {
  expect(generateD2(fakeDiagram)).toMatchSnapshot()
})

test('generate D2 - fakeDiagram2', () => {
  expect(generateD2(fakeDiagram2)).toMatchSnapshot()
})

test('generate D2 - fakeComputedView 3 Levels', () => {
  expect(generateD2(fakeComputedView3Levels)).toMatchSnapshot()
})
