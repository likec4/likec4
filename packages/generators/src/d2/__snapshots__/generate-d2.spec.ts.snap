// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate D2 - fakeComputedView 3 Levels 1`] = `
"Amazon: {
  label: "amazon"
}
Cloud: {
  label: "cloud"

  Backend: {
    label: "backend"

    Graphql: {
      label: "graphql"
    }
    Storage: {
      label: "storage"
    }
  }
  FrontendAdminPanel: {
    label: "adminPanel"
  }
  FrontendDashboard: {
    label: "dashboard"
  }
}
Customer: {
  label: "customer"
}
Support: {
  label: "support"
}

Cloud.FrontendDashboard -> Cloud.Backend.Graphql
Cloud.FrontendAdminPanel -> Cloud.Backend.Graphql
Cloud.Backend.Storage -> Amazon
Cloud.Backend.Graphql -> Cloud.Backend.Storage
Support -> Cloud.FrontendAdminPanel
Customer -> Cloud.FrontendDashboard
"
`;

exports[`generate D2 - fakeDiagram 1`] = `
"Amazon: {
  label: "amazon"
}
Cloud: {
  label: "cloud"

  Backend: {
    label: "backend"
  }
  Frontend: {
    label: "frontend"
  }
}
Customer: {
  label: "customer"
}
Support: {
  label: "support"
}

Cloud.Frontend -> Cloud.Backend: requests
Cloud.Backend -> Amazon
Support -> Cloud.Frontend
Customer -> Cloud.Frontend: opens
"
`;

exports[`generate D2 - fakeDiagram2 1`] = `
"Client: {
  label: "client"
}
SystemBackend: {
  label: "backend"
}
SystemFrontend: {
  label: "frontend"
}

SystemFrontend -> SystemBackend: requests
Client -> SystemFrontend: opens
"
`;
